File: parser/src/validators/data_validator.py
"""Data validation utilities for scraped Goodreads data.

Validates ratings, fields, and data quality per FR-013.
Per Constitution Principle VI: Data Quality & Validation.
"""

from datetime import datetime
from typing import Any

from src.exceptions import ValidationError


def validate_rating(rating: int | None, field_name: str = "rating") -> int | None:
    """Validate a 1-5 star rating.

    Args:
        rating: Rating value to validate (1-5 or None)
        field_name: Name of field for error messages

    Returns:
        Validated rating or None

    Raises:
        ValidationError: If rating is not in valid 1-5 range

    Examples:
        >>> validate_rating(5)
        5
        >>> validate_rating(None)
        None
        >>> validate_rating(6)  # doctest: +SKIP
        ValidationError: rating must be between 1 and 5, got: 6
    """
    if rating is None:
        return None

    if not isinstance(rating, int):
        raise ValidationError(
            f"{field_name} must be an integer, got: {type(rating).__name__}"
        )

    if not (1 <= rating <= 5):
        raise ValidationError(
            f"{field_name} must be between 1 and 5, got: {rating}"
        )

    return rating


def sanitize_text(text: str | None, max_length: int | None = None) -> str | None:
    """Sanitize text by stripping whitespace and optionally truncating.

    Args:
        text: Text to sanitize
        max_length: Maximum length (truncated if exceeded, None for no limit)

    Returns:
        Sanitized text or None if input was None/empty

    Examples:
        >>> sanitize_text("  Hello World  ")
        'Hello World'
        >>> sanitize_text("Too long text", max_length=8)
        'Too long'
        >>> sanitize_text(None)
        None
    """
    if text is None:
        return None

    text = text.strip()

    if not text:
        return None

    if max_length is not None and len(text) > max_length:
        text = text[:max_length]

    return text


def validate_isbn(isbn: str | None) -> str | None:
    """Validate ISBN format (basic check).

    Note: Full ISBN validation with checksum is handled by Pydantic ISBN type.
    This is a basic sanity check for scraped data.

    Args:
        isbn: ISBN string to validate

    Returns:
        Normalized ISBN (digits only) or None

    Examples:
        >>> validate_isbn("978-0525478812")
        '9780525478812'
        >>> validate_isbn("0-525-47881-2")
        '052547881 2'
        >>> validate_isbn(None)
        None
    """
    if isbn is None:
        return None

    # Remove common separators
    isbn_clean = isbn.replace("-", "").replace(" ", "")

    if not isbn_clean:
        return None

    # ISBN-10 should be 10 digits, ISBN-13 should be 13 digits
    if len(isbn_clean) not in (10, 13):
        return None

    # Should be mostly digits (allowing 'X' for ISBN-10 check digit)
    if not all(c.isdigit() or c.upper() == 'X' for c in isbn_clean):
        return None

    return isbn


def validate_publication_year(year: int | None) -> int | None:
    """Validate publication year is reasonable.

    Args:
        year: Publication year to validate

    Returns:
        Validated year or None

    Raises:
        ValidationError: If year is outside reasonable range (1000-2100)

    Examples:
        >>> validate_publication_year(2020)
        2020
        >>> validate_publication_year(None)
        None
        >>> validate_publication_year(999)  # doctest: +SKIP
        ValidationError: publication_year must be between 1000 and 2100, got: 999
    """
    if year is None:
        return None

    if not isinstance(year, int):
        raise ValidationError(
            f"publication_year must be an integer, got: {type(year).__name__}"
        )

    if not (1000 <= year <= 2100):
        raise ValidationError(
            f"publication_year must be between 1000 and 2100, got: {year}"
        )

    return year


def validate_page_count(pages: int | None) -> int | None:
    """Validate page count is positive.

    Args:
        pages: Page count to validate

    Returns:
        Validated page count or None

    Raises:
        ValidationError: If page count is not positive

    Examples:
        >>> validate_page_count(350)
        350
        >>> validate_page_count(None)
        None
        >>> validate_page_count(0)  # doctest: +SKIP
        ValidationError: page_count must be positive, got: 0
    """
    if pages is None:
        return None

    if not isinstance(pages, int):
        raise ValidationError(
            f"page_count must be an integer, got: {type(pages).__name__}"
        )

    if pages < 1:
        raise ValidationError(
            f"page_count must be positive, got: {pages}"
        )

    return pages


def normalize_genres(genres: list[str]) -> list[str]:
    """Normalize genre list: lowercase, deduplicate, limit length.

    Args:
        genres: List of genre strings

    Returns:
        Normalized list of genres (max 50, deduplicated, lowercased)

    Examples:
        >>> normalize_genres(["Fiction", "MYSTERY", "fiction"])
        ['fiction', 'mystery']
        >>> normalize_genres([])
        []
    """
    if not genres:
        return []

    normalized = []
    seen = set()

    for genre in genres[:50]:  # Limit to max 50 genres
        if not isinstance(genre, str):
            continue

        genre_clean = genre.strip().lower()[:50]  # Max 50 chars per genre

        if genre_clean and genre_clean not in seen:
            normalized.append(genre_clean)
            seen.add(genre_clean)

    return normalized


def parse_iso_date(date_str: str | None) -> datetime | None:
    """Parse ISO 8601 date string to datetime.

    Args:
        date_str: ISO 8601 formatted date string

    Returns:
        Parsed datetime object or None

    Examples:
        >>> parse_iso_date("2024-03-15T10:30:00Z")  # doctest: +SKIP
        datetime.datetime(2024, 3, 15, 10, 30)
        >>> parse_iso_date(None)
        None
    """
    if not date_str:
        return None

    try:
        # Handle various ISO 8601 formats
        # Remove 'Z' suffix if present
        if date_str.endswith('Z'):
            date_str = date_str[:-1] + '+00:00'

        return datetime.fromisoformat(date_str)
    except (ValueError, AttributeError):
        return None


def validate_date_ordering(
    date_added: datetime | None,
    date_started: datetime | None,
    date_finished: datetime | None
) -> tuple[bool, str | None]:
    """Validate date ordering: added <= started <= finished.

    Args:
        date_added: When book was added to library
        date_started: When reading started
        date_finished: When reading finished

    Returns:
        Tuple of (is_valid, error_message)
        is_valid: True if dates are in correct order
        error_message: Description of validation failure, None if valid

    Examples:
        >>> from datetime import datetime
        >>> validate_date_ordering(
        ...     datetime(2024, 1, 1),
        ...     datetime(2024, 1, 10),
        ...     datetime(2024, 1, 20)
        ... )
        (True, None)
    """
    # Hard validation: started must be <= finished
    if date_started and date_finished:
        if date_started > date_finished:
            return (
                False,
                f"date_started ({date_started}) cannot be after date_finished ({date_finished})"
            )

    # Soft validation: added should be <= started (warning only, not error)
    # This is handled as a warning in the scraper logging
    return (True, None)

File: parser/src/validators/data-validator.ts
export class DataValidator {
  /**
   * Validate rating is between 1-5
   */
  static validateRating(rating: number | null | undefined): boolean {
    if (rating === null || rating === undefined) {
      return true; // null ratings are valid
    }
    return rating >= 1 && rating <= 5;
  }

  /**
   * Sanitize text by trimming whitespace
   */
  static sanitizeText(text: string | null | undefined): string | null {
    if (!text) return null;
    const trimmed = text.trim();
    return trimmed.length > 0 ? trimmed : null;
  }

  /**
   * Validate ISBN-10 or ISBN-13
   */
  static validateIsbn(isbn: string | null | undefined): boolean {
    if (!isbn) return true; // null ISBNs are valid

    // Remove hyphens and spaces
    const cleaned = isbn.replace(/[-\s]/g, '');

    // Check length
    if (cleaned.length !== 10 && cleaned.length !== 13) {
      return false;
    }

    // Check if all characters are digits (except last char in ISBN-10 can be X)
    if (cleaned.length === 10) {
      return /^\d{9}[\dX]$/i.test(cleaned);
    } else {
      return /^\d{13}$/.test(cleaned);
    }
  }

  /**
   * Validate publication year is reasonable
   */
  static validatePublicationYear(year: number | null | undefined): boolean {
    if (year === null || year === undefined) return true;
    const currentYear = new Date().getFullYear();
    return year >= 1000 && year <= currentYear + 5; // Allow a few years ahead for pre-orders
  }

  /**
   * Validate page count is positive
   */
  static validatePageCount(count: number | null | undefined): boolean {
    if (count === null || count === undefined) return true;
    return count > 0 && Number.isInteger(count);
  }

  /**
   * Normalize genres (lowercase, deduplicate, trim)
   */
  static normalizeGenres(genres: string[]): string[] {
    if (!genres || !Array.isArray(genres)) return [];

    return genres
      .map(g => g.toLowerCase().trim())
      .filter((g, index, arr) => g.length > 0 && arr.indexOf(g) === index) // deduplicate
      .slice(0, 50); // Limit to 50 genres
  }

  /**
   * Parse ISO date string
   */
  static parseIsoDate(dateStr: string | null | undefined): Date | null {
    if (!dateStr) return null;

    try {
      const date = new Date(dateStr);
      return isNaN(date.getTime()) ? null : date;
    } catch {
      return null;
    }
  }

  /**
   * Validate date ordering (start <= end)
   */
  static validateDateOrdering(
    startDate: string | null | undefined,
    endDate: string | null | undefined
  ): boolean {
    if (!startDate || !endDate) return true;

    const start = this.parseIsoDate(startDate);
    const end = this.parseIsoDate(endDate);

    if (!start || !end) return true;

    return start <= end;
  }
}
